R/tfb-wavelet-utils.R

```diff
@@ -0,0 +1,166 @@
+remove_slope <- function(x, y) {
+  last_element <- length(x)
+  slope <- (y[last_element] - y[1]) / (x[last_element] - x[1])
+  intercept <- y[1] - slope * x[1]
+  f <- intercept + slope * x
+  y_desloped <- y - f
+  y_desloped <- structure(f,
+                          intercept = intercept,
+                          slope = slope)
+  y_desloped
+}
+
+
+interpolate_arg <- function(arg_list) {
+  n <- vapply(arg_list, length, numeric(1))
+  closest_power <- round(log2(n))
+  n_diff <- 2^closest_power - n
+  
+  interp_index <- map2(arg_list, 
+                       n_diff,
+                       function(x, y) {
+                         slope <- (max(x) - min(x)) / (length(x) + y - 1)
+                         seq(min(x), max(x), by = slope)
+                       })
+  unlist(interp_index)
+}
+
+
+
+fit_wavelet <- function(data, Z, penalized, glmnet_args) {
+  eval_list <- split(data$value, data$id)
+  arg_list <- unique(data$arg)
+  
+  Z <- cbind(Z, arg_list)
+  Z <- scale(Z, center = FALSE)
+  
+  if (!penalized) {
+    coefs <- map(eval_list, 
+                  function(x) {
+                    # Least squares fit, directly computed
+                    unname(lsfit(Z, x, intercept = TRUE)$coef)})
+  } else {
+    coefs <- map(eval_list, 
+                  function(y) {
+                    temp_model <- do.call(glmnet::cv.glmnet, c(list(Z, y), 
+                                                               glmnet_args))
+                    as.numeric(coefficients(temp_model))
+                  }
+    )
+  }
+  coefs
+}
+
+
+fit_wavelet_irr <- function(data, Z, penalized, glmnet_args) {
+  eval_list <- split(data$value, data$id)
+  arg_list <- split(data$arg, data$id)
+  
+  if (!penalized) {
+    coefs <- map2(arg_list, eval_list,  
+                  function(x, y) {
+                    Z <- cbind(Z, x)
```

Comment:
this seems off--
1. in order for tf_evaluate to work precisely all the estimated coefficients need to use (a subset of rows) of the SAME Z.
2. here, this will be different since x will be different and so `scale` will change the values of that column between different functions.
3. if these are irregular data, their fit will be based on different subsets of rows of Z corresponding to the different grids on which they are observed -- compare to https://github.com/tidyfun/tidyfun/blob/2fdced5fd6d11fb3e3f572de319bd1abaf9916ea/R/tfb-spline-utils.R#L93-L97

Reply:
@SvenLorenz any progress on this yet?

Reply:
Yes, i think this works now. Although default evaluations are still a bit clunky, because every curve is evaluated at every arg that is available.